File: agentic-patterns/go/flows/tool_calling.go
package flows

import (
	"context"
	"fmt"

	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/core"
	"github.com/firebase/genkit/go/genkit"
)

type ToolCallingRequest struct {
	Prompt string `json:"prompt"`
}

type GetWeatherRequest struct {
	Location string `json:"location"`
}

func DefineToolCallingFlow(g *genkit.Genkit) *core.Flow[*ToolCallingRequest, string, struct{}] {
	getWeather := genkit.DefineTool(g,
		"getWeather",
		"Get the current weather in a given location.",
		func(ctx *ai.ToolContext, req *GetWeatherRequest) (string, error) {
			// In a real app, you would call a weather API here.
			return fmt.Sprintf("The weather in %s is 75°F and sunny.", req.Location), nil
		},
	)

	return genkit.DefineFlow(g, "toolCallingFlow",
		func(ctx context.Context, req *ToolCallingRequest) (string, error) {
			response, err := genkit.Generate(ctx, g,
				ai.WithPrompt(req.Prompt, nil),
				ai.WithTools(getWeather),
			)
			if err != nil {
				return "", err
			}
			return response.Text(), nil
		},
	)
}

File: agentic-patterns/js/src/tool-calling.ts
import { z } from 'genkit';
import { ai } from './genkit.js';

// Define a tool that can be called by the LLM
const getWeather = ai.defineTool(
  {
    name: 'getWeather',
    description: 'Get the current weather in a given location.',
    inputSchema: z.object({ location: z.string() }),
    outputSchema: z.string(),
  },
  async ({ location }) => {
    // In a real app, you would call a weather API here.
    return `The weather in ${location} is 75°F and sunny.`;
  }
);

export const toolCallingFlow = ai.defineFlow(
  {
    name: 'toolCallingFlow',
    inputSchema: z.object({ prompt: z.string() }),
    outputSchema: z.string(),
  },
  async ({ prompt }) => {
    const response = await ai.generate({
      prompt: prompt,
      tools: [getWeather],
    });

    return response.text;
  }
);
